### PR TITLE
feat: add region migration procedure skeleton

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add the skeleton of region migration procedure.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/meta-srv/src/procedure/region_migration.rs [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+impl PersistentContext {
+    pub fn lock_key(&self) -> String {
+        let table_name = &self.table_name;
+
+        common_catalog::format_full_table_name(
+            &table_name.catalog_name,
+            &table_name.schema_name,
+            &table_name.table_name,
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/meta-srv/src/procedure/region_migration.rs:59]: I double-checked the DDL procedures. If we only acquire the region granularity lock in the region migration procedure, we must acquire all region locks during the alter procedure.

File: src/meta-srv/src/procedure/region_migration.rs [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PersistentContext {
+    /// The [Peer] of migration source.
+    src_peer: Peer,
+    /// The [Peer] of migration destination.
+    dst_peer: Option<Peer>,
+    /// Closes the migrated region on the `src_peer`.
+    close_migrated_region: bool,
+    /// The [TableName] of migration region.
+    table_name: TableName,
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/meta-srv/src/procedure/region_migration.rs]: Removed in the [commit](https://github.com/WenyXu/greptimedb/tree/3b4d21b134561c630946e939ba91b009003df13b)

File: src/meta-srv/src/procedure/region_migration.rs [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+///
+/// **Notes: Stores with too large data in the context might incur replication overhead.**
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PersistentContext {
+    /// The [Peer] of migration source.
+    src_peer: Peer,
+    /// The [Peer] of migration destination.
+    dst_peer: Option<Peer>,
+    /// Closes the migrated region on the `src_peer`.
+    close_migrated_region: bool,
```
[the remark this answers is not in the document]
Comment (MichaelScofield) [reply src/meta-srv/src/procedure/region_migration.rs:44]: Shouldn't the migrated regions always be closed in the `from_peer`?

File: src/meta-srv/src/procedure/region_migration.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/meta-srv/src/procedure/region_migration.rs]: I logged this issue in #2750. It relates to multiple procedures.

File: src/meta-srv/src/procedure/region_migration.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/meta-srv/src/procedure/region_migration.rs]: > Shouldn't the migrated regions always be closed in the `from_peer`?

Yes, I removed it now.

**Comment by WenyXu** [comment]
@MichaelScofield @fengjiachun PTAL

**Review by MichaelScofield — APPROVED** [review]

File: src/meta-srv/src/procedure/region_migration.rs [hunk 4]
```diff
@@ -38,25 +37,21 @@ use crate::error::{Error, Result};
...
-        let table_name = &self.table_name;
-
-        common_catalog::format_full_table_name(
-            &table_name.catalog_name,
-            &table_name.schema_name,
-            &table_name.table_name,
+        format!(
+            "{}/{}",
+            self.region_id.table_id(),
+            self.region_id.region_number()
```
Comment (fengjiachun) [inline src/meta-srv/src/procedure/region_migration.rs:52]: Region migration needs to obtain the same lock as region failover.

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**